### PR TITLE
[docs] Add "currency type" to the lexicon

### PR DESCRIPTION
**Pull request by philipturner**
I don't understand why it says "currency", which means money. There might be some Swift data type for money, but I haven't heard of it yet. Also, the explanation misses that the entire document just covers built-in Swift types. Did you mean "concurrency"?

### CODE REVIEW

**Review by CodaFi — CHANGES_REQUESTED** [review]
There is a difference here!

The compiler understands `Builtin` types like `Builtin.NativeObject`, and it understands much higher-level common user types like `String`. The types in the standard library are not considered builtins - they're built atop builtins more often than not. The document linked here describes optimization opportunities exposed by annotations and structures attached these higher-level types and associated operations.

"Currency" (often also "vocabulary") types are types that one reaches for in most common usages of the language. Do you want to manipulate a string - pick `String`, not `UnsafeMutableBufferPointer<CChar>`. Do you want a pack of unique elements - pick `Set<T>`!

**Comment by philipturner** [comment]
Then could the documentation clarify what "currency" means? I didn't see it in `Lexicon.md`, and from the point of view of someone who doesn't know what it means, it looks like a misspelling of "Concurrency".

**Comment by WowbaggersLiquidLunch** [comment]
also relevant answer on the forums: https://forums.swift.org/t/what-does-currency-type-mean/41065/2

**Comment by philipturner** [comment]
The definition for currency still needs to be refined, could anyone help me here?

File: docs/Lexicon.md [hunk 1]
```diff
@@ -137,6 +137,10 @@ the AST level. See also [witness table](#witness-table).
 An edge in a control flow graph where the destination has multiple predecessors
 and the source has multiple successors.
 
+## currency
```
Comment (CodaFi) [inline docs/Lexicon.md:140]:
```suggestion
## currency type
```

File: docs/Lexicon.md [hunk 2]
```diff
@@ -137,6 +137,10 @@ the AST level. See also [witness table](#witness-table).
...
+
+Aa type that's meant to be commonly passed around and stored, like `Array`, as opposed to a type that's useful for temporary/internal purposes but which you wouldn't normally use in an external interface, like `ArraySlice`. Having broad agreement about the currency type you use for a particular kind of data (e.g. using `Array` to pass around sequential collections) generally makes the whole ecosystem better by reducing artificial barriers to passing data from one system to another, and it gives algorithm writers an obvious target to ensure they optimize for. That's where the analogy to currency comes from: agreeing on a currency type improves the flow of information in a program in some of the same ways that agreeing on a currency improves the flow of trade in an economy.
```
Comment (CodaFi) [inline docs/Lexicon.md:142]: Nit: Could you reflow this text to 80(-ish) columns?

File: docs/README.md [hunk 3]
```diff
@@ -131,7 +131,7 @@ documentation, please create a thread on the Swift forums under the
     Describes the design of the optimizer pipeline.
   - [HighLevelSILOptimizations.rst](/docs/HighLevelSILOptimizations.rst):
     Describes how the optimizer understands the semantics of high-level
-    operations on currency data types and optimizes accordingly.
+    operations on [currency](/docs/Lexicon.md#currency) data types and optimizes accordingly.
```
Comment (CodaFi) [inline docs/README.md:134]: This link needs to be updated as well.

Comment (philipturner) [reply docs/README.md:134]:
```suggestion
    operations on [currency](/docs/Lexicon.md#currency-types) data types and optimizes accordingly.
```

**Review by CodaFi — APPROVED** [review]
Thanks

⛵

**Comment by CodaFi** [comment]
Note to self: manually squash this

@swift-ci smoke test

File: docs/README.md [hunk 4]
```diff
@@ -131,7 +131,7 @@ documentation, please create a thread on the Swift forums under the
     Describes the design of the optimizer pipeline.
   - [HighLevelSILOptimizations.rst](/docs/HighLevelSILOptimizations.rst):
     Describes how the optimizer understands the semantics of high-level
-    operations on currency data types and optimizes accordingly.
+    operations on [currency](/docs/Lexicon.md#currency-types) data types and optimizes accordingly.
```
Comment (xwu) [inline docs/README.md:134]: Link's not quite correct.

```suggestion
    operations on [currency](/docs/Lexicon.md#currency-type) data types and 
    optimizes accordingly.
```

**Comment by philipturner** [comment]
@CodaFi you may need to retest this because of a one-character change.

**Comment by xwu** [comment]
@swift-ci smoke test

**Pull request merged**